NYPLAudiobookToolkit/Network/OpenAccessDownloadTask.swift

```diff
@@ -188,6 +202,53 @@ final class OpenAccessDownloadTask: DownloadTask {
         }
         return hash
     }
+  
+    // MARK: - Download Timer
+  
+    private func startDownloadTimer() {
+        serialQueue.async {
+            /// These value should be reset every time we start fetching,
+            /// so that we have the right time in case of retrying download
+            self.fetchStartTime = Date()
+            self.downloadTimeLimit = 30.0
+            
+            self.downloadTimer = NYPLRepeatingTimer(interval: .seconds(30),
+                                                    queue: self.serialQueue,
+                                                    handler: { [weak self] in
+                guard let self = self,
+                      let startTime = self.fetchStartTime else {
+                    return
+                }
+              
+                let elapsedTime = Date().timeIntervalSince(startTime)
+                if elapsedTime >= self.downloadTimeLimit {
+                    self.delegate?.downloadTaskExceededTimeLimit(self,
+                                                                 elapsedTime: elapsedTime)
+                    self.updateDownloadTimeLimit()
+                }
+            })
+        }
+    }
+  
+    fileprivate func removeDownloadTimer() {
+        serialQueue.async {
+            self.downloadTimer = nil
+        }
+    }
+  
+    /// We call the delegate when the download task has not completed
+    /// at the 30 seconds mark and 3 minutes (180 seconds) mark.
+    /// Therefore, we update the time limit every time the delegate is called and
+    /// remove the timer when we are done with it.
+    private func updateDownloadTimeLimit() {
+        serialQueue.async {
+            if self.downloadTimeLimit == 30.0 {
+                self.downloadTimeLimit = 180.0
+            } else if self.downloadTimeLimit == 180.0 {
+                self.removeDownloadTimer()
```

Comment:
if we never get to this limit, when are we removing the timer?

Reply:
Whenever the download task failed or completed, or when we cancel it.